File: YangTuanAnh/lesson_1/Happy Number.cpp
class Solution {
public:
    int happy(int n) {
        int ans = 0;
        while (n) {
            ans+=(n%10)*(n%10);
            n/=10;
        }
        return ans;
    }
    bool isHappy(int n) {
        int slow = n, fast = n;
        while (fast!=1) {
            slow = happy(slow);
            fast = happy(fast);
            fast = happy(fast);
            if (slow==fast && slow!=1) return false;
        }
        return true;
    }
};

File: Homework/Chau_Long/lesson_1/happy_number.py
class Solution:
    def isHappy(self, n: int) -> bool:
        def get_next_n(num: int) -> int:
            res = 0
            while num:
                num, digit = divmod(num, 10)
                res += digit ** 2

            return res

        slow = n
        fast = get_next_n(n)
        while fast != 1 and fast != slow:
            slow = get_next_n(slow)
            fast = get_next_n(get_next_n(fast))

        return fast == 1

if __name__ == "__main__":
    sol = Solution()
    print(sol.isHappy(19) == True)
    print(sol.isHappy(2) == False)